### DISCUSSION TITLE
How to write a middleware?

### DISCUSSION THREAD

**Discussion by budarin** [pmndrs/zustand#1770] [Q&A] [answered]
I want to write my middleware, but I haven't found documentation anywhere describing the API and how it works

Those middleware examples that are in the project are very difficult to understand and their API and typing are very difficult for a beginner to understand

**Comment by dai-shi** [comment, marked as answer]
Interestingly, there's nothing in zustand code about middleware.
It's just an enhancer to `createState` function. It's a function you pass to `create`.
In that sense, it's a userland solution and there's no API per se.

So, in JS, it's fairly easy, like the example: https://github.com/pmndrs/zustand#middleware
However, zustand do support middleware through TypeScript, and it's very difficult.

If you find existing middleware too difficult, you don't bother about middleware.
As app developers, you should be able to do things after creating a store and usually it's easier.
If you want to develop a library with middleware, you would need to learn it anyway from code.

**Comment by budarin** [comment]
Hi! 

Thanks for the response.
I will try to write middleware and if I encounter problems, they will ask questions here

**Comment by budarin** [comment]
I have the [redux-business-logic-middleare](https://www.npmjs.com/package/@budarin/redux-business-logic-middleare)
I want to have the same middleware in Zustand for using with redux middleware

I need your help to writing it

```ts
const actionHandlers = new Map();

export function onAction(actionId: string, handler: (set, get, api, action) => void) {
    actionHandlers.set(actionId, handler);
}

export function offAction(actionId: string) {
    actionHandlers.delete(actionId);
}


export const businessLogic = (config) => (set, get, api) => {
    const initialState = config(set, get, api);
    const originalDispath = api.dispatch;

    api.dispatch = (action: any) => {
        try {
            const handler = actionHandlers.get(action.type);

            if (!handler) {
                return originalDispath(action);
            }

            return handler(set, get, api, action);
        } finally {
        }
    };
    return initialState;
};
```

I need to type the interface for a custom business logic handler:
```ts
(set, get, api, action) => void
```

**Reply by dai-shi** [reply]
Does it work fine in JS, and is it just TS issue?

Please check out the docs if you haven't: https://github.com/pmndrs/zustand/blob/513b22283eadae6fe2a5ad8a41487916b47a78b1/docs/guides/typescript.md?plain=1#L210

**Comment by budarin** [comment]
It's typescript issue for typing handlers in app code, not for middleware

**Reply by dai-shi** [reply]
You want to create a reproduction with TypeScript playground https://tsplay.dev/wj0E2m to get help from someone.

**Comment by walunjakarsachinedu** [comment]
I have implemented a type-safe middleware for Zustand that wraps store actions to execute hooks before and/or after the action runs. The middleware provides access to set, get, and the action name, enabling enhanced control over state updates. I wrote this to help anyone looking for the simplest middleware integration.

```typescript
import { StateCreator } from "zustand";

type StoreValueSetter<State> = {
  (
    partial:
      | State
      | Partial<State>
      | ((state: State) => State | Partial<State>),
    replace?: false
  ): void;
  (state: State | ((state: State) => State), replace: true): void;
};

export type Middleware<State> = (
  action: string,
  set: StoreValueSetter<State>,
  get: () => State
) => void;

export default function applyMiddleware<State extends object>({
  store,
  beforeMiddlware,
  afterMiddlware,
}: {
  store: StateCreator<State, [], []>;
  beforeMiddlware?: Middleware<State>;
  afterMiddlware?: Middleware<State>;
}): StateCreator<State, [], []> {
  return (set, get, storeApi) => {
    /** Wrap each store action with middleware: beforeMiddleware runs before, afterMiddleware runs after the action executes. */
    function storeWrapper<P extends unknown[], R>(
      fn: (...args: P) => R // The original function
    ): (...args: P) => R {
      return (...args: P): R => {
        beforeMiddlware?.(fn.name, set, get);
        const result = fn(...args);
        afterMiddlware?.(fn.name, set, get);
        return result;
      };
    }

    const initializedStore = store(set, get, storeApi);

    for (const key in initializedStore) {
      if (Object.prototype.hasOwnProperty.call(initializedStore, key)) {
        if (typeof initializedStore[key] === "function") {
          initializedStore[key] = storeWrapper(
            initializedStore[key] as (...args: unknown[]) => unknown
          ) as State[Extract<keyof State, string>];
        }
      }
    }
    return initializedStore;
  };
}
```
